Terraform Docs (#3174)
hashicorp/web-unified-docs: content/terraform-policy/v0.2.x (beta)/docs/policy/reference/cli/validate.mdx

---
page_title: '`tfpolicy validate` command reference'
description: >-
  Reference for the `tfpolicy validate` command to validate policy syntax.
# START AUTO GENERATED METADATA, DO NOT EDIT
created_at: 2026-07-14T11:44:47-05:00
last_modified: 2026-07-14T19:01:23.000Z
# END AUTO GENERATED METADATA
---

# `tfpolicy validate`

The `tfpolicy validate` command validates the syntax of your policy files.

@include "beta.mdx"

## Usage

```shell-session
$ tfpolicy validate --policies=</path/to/policies>
```

## Description

> Once you have written your `.policytest.hcl` file, validate their syntax with the `tfpolicy validate` command. Similar to the `terraform validate` command, `tfpolicy validate` reports whether your policies are syntactically valid and internally consistent.

Reviewer: `validate` only checks `.policy.hcl` files - test files are validated by `tfpolicy test`, so this should say `.policy.hcl`.

> Once you have written your `.policytest.hcl` file, validate their syntax with the `tfpolicy validate` command. Similar to the `terraform validate` command, `tfpolicy validate` reports whether your policies are syntactically valid and internally consistent.

Author: ```suggestion
Once you have written your `.policy.hcl` files, validate their syntax with the `tfpolicy validate` command. Similar to the `terraform validate` command, `tfpolicy validate` reports whether your policies are syntactically valid and internally consistent.
```

> If you define a `required_providers` block for your policies, Terraform policy will validate your provider, resource, and data source policy attributes, as well as the arguments to the [`core::getdatasource`](/terraform/policy/reference/functions/getdatasource) and [`core::getresources`](/terraform/policy/reference/functions/getresources) functions against the schemas of the providers specified. Policies that match multiple resource types with `*` must validate against each matching resource or data source type that matches the expression. For more information, refer to [`required_providers`](/terraform/policy/reference/policy/policy#required_providers).

Author: ```suggestion
The `required_providers` block is mandatory for policy files which contain resource or provider policies. Terraform policy will use this configuration to validate your provider, resource, and data source policy attributes, as well as the arguments to the [`core::getdatasource`](/terraform/policy/reference/functions/getdatasource) and [`core::getresources`](/terraform/policy/reference/functions/getresources) functions against the schemas of the providers specified. Policies that match multiple resource types with `*` must validate against each matching resource or data source type that matches the expression. For more information, refer to [`required_providers`](/terraform/policy/reference/policy/policy#required_providers).
```

> The `required_providers` block is mandatory for policy files which contain resource or provider policies. Terraform policy will use this configuration to validate your provider, resource, and data source policy attributes, as well as the arguments to the [`core::getdatasource`](/terraform/policy/reference/functions/getdatasource) and [`core::getresources`](/terraform/policy/reference/functions/getresources) functions against the schemas of the providers specified. Policies that match multiple resource types with `*` must validate against each matching resource or data source type that matches the expression. For more information, refer to [`required_providers`](/terraform/policy/reference/policy/policy#required_providers).

Reviewer 2: ```suggestion
The `required_providers` block is mandatory for policy files that contain resource or provider policies. Terraform policy applies this configuration when validating your provider, resource, and data source policy attributes, as well as the arguments to the [`core::getdatasource`](/terraform/policy/reference/functions/getdatasource) and [`core::getresources`](/terraform/policy/reference/functions/getresources) functions against the schemas of the providers specified. Policies that match multiple resource types with `*` must validate against each matching resource or data source type that matches the expression. For more information, refer to [`required_providers`](/terraform/policy/reference/policy/policy#required_providers).
```

> The `required_providers` block is mandatory for policy files which contain resource or provider policies. Terraform policy will use this configuration to validate your provider, resource, and data source policy attributes, as well as the arguments to the [`core::getdatasource`](/terraform/policy/reference/functions/getdatasource) and [`core::getresources`](/terraform/policy/reference/functions/getresources) functions against the schemas of the providers specified. Policies that match multiple resource types with `*` must validate against each matching resource or data source type that matches the expression. For more information, refer to [`required_providers`](/terraform/policy/reference/policy/policy#required_providers).

Reviewer 2: ```suggestion
The `required_providers` block is mandatory for policy files which contain resource or provider policies. Terraform policy uses this configuration to validate your provider, resource, and data source policy attributes, as well as the arguments to the [`core::getdatasource`](/terraform/policy/reference/functions/getdatasource) and [`core::getresources`](/terraform/policy/reference/functions/getresources) functions against the schemas of the providers specified. Policies that match multiple resource types with `*` must validate against each matching resource or data source type that matches the expression. For more information, refer to [`required_providers`](/terraform/policy/reference/policy/policy#required_providers).
```

## Available options

The `tfpolicy validate` command supports the following options.

| Option | Description | Default | Required |
| ------ | ----------- | ------- | -------- |
| `--policies=</path/to/policies>` | Validate the syntax of all policies found in the specified path. | Current directory | No |
| `--evaluation-stage=<stage>` | Validate policies for the specified evaluation stage: `"setup"` or `"plan"`. | `"plan"` | No |

## Exit code

> The `tfpolicy validate` command will exit with a non-zero exit code when validation fails, or an exit code of `0` if validation is successful.

Reviewer 2: ```suggestion
The `tfpolicy validate` command exits with either a non-zero exit code when any validation fails or an exit code of `0` if validation is successful.
```

Once you have written your `.policytest.hcl` file, validate their syntax with the `tfpolicy validate` command. Similar to the `terraform validate` command, `tfpolicy validate` reports whether your policies are syntactically valid and internally consistent.

> If you define a `required_providers` block for your policies, Terraform policy will validate your provider, resource, and data source policy attributes, as well as the arguments to the [`core::getdatasource`](/terraform/policy/reference/functions/getdatasource) and [`core::getresources`](/terraform/policy/reference/functions/getresources) functions against the schemas of the providers specified. Policies that match multiple resource types with `*` must validate against each matching resource or data source type that matches the expression. For more information, refer to [`required_providers`](/terraform/policy/reference/policy/policy#required_providers).

Reviewer 3: I believe the required_providers block is mandatory. cc: @viveksinghgit95 to confirm

Reviewer: Confirmed, it's required whenever a file has a `resource_policy` or `provider_policy` (`module_policy` is exempt).

Author: Updated here, and also added notes in the relevant places in the [usage docs](https://github.com/hashicorp/web-unified-docs/pull/3174/changes#r3854850355) and [reference page](https://github.com/hashicorp/web-unified-docs/pull/3174/changes#r3854819983).
